### PR TITLE
Checkout Commit 1

### PR DESCRIPTION
**Pull request by mariusraht2**
New Feature "Checkout Commit"; Part 1
Related to #3040 - Revert to any single commit

### Change Log
**zcl_abapgit_git_porcelain**
- Methods `pull` and `push` consider retrieving git objects by any commit hash, not only the branches one

**zcl_abapgit_git_transport**
- Added some attributes and related logic to `upload_pack` to be able to retrieve commits not only by branch commit and be able to specify the deepen level for the commit hierarchie to be retrieved

**zif_abapgit_persistence**
- Added field commit_sha1 as representation of attribute commit_sha1 of online repos to be able to use them in method `set` of class repo

**zcl_abapgit_merge**
- Adjusted call of method `upload_pack` to new method signature

**zcl_abapgit_repo**
- Made `get_local_checksums` public
- Now it's possible to set commit_sha1 for repo

**zcl_abapgit_repo_online**
- `set_branch_name`: renamed` iv_branch_name` to `iv_name`
- Added getter and setter for commit hash

**zif_abapgit_definitions**
- Added type definition for git commit and a list of the same structure type

### CODE REVIEW

File: src/git/zcl_abapgit_git_porcelain.clas.abap [hunk 1]
```diff
@@ -4,31 +4,32 @@ CLASS zcl_abapgit_git_porcelain DEFINITION
...
+             objects TYPE zif_abapgit_definitions=>ty_objects_tt,
+             branch  TYPE zif_abapgit_definitions=>ty_sha1,
+             commit  TYPE zif_abapgit_definitions=>ty_sha1,
+           END OF ty_pull_result .
+    TYPES: BEGIN OF ty_push_result,
+             new_files     TYPE zif_abapgit_definitions=>ty_files_tt,
+             branch        TYPE zif_abapgit_definitions=>ty_sha1,
+             updated_files TYPE zif_abapgit_definitions=>ty_file_signatures_tt,
+             new_objects   TYPE zif_abapgit_definitions=>ty_objects_tt,
+           END OF ty_push_result .
```
Comment (larshp) [inline src/git/zcl_abapgit_git_porcelain.clas.abap]: editing the class via SE24 will format this back to what it was before, makes it more difficult to review

File: src/git/zcl_abapgit_git_porcelain.clas.abap [hunk 2]
```diff
@@ -4,31 +4,32 @@ CLASS zcl_abapgit_git_porcelain DEFINITION
...
-        new_files     TYPE zif_abapgit_definitions=>ty_files_tt,
-        branch        TYPE zif_abapgit_definitions=>ty_sha1,
-        updated_files TYPE zif_abapgit_definitions=>ty_file_signatures_tt,
-        new_objects   TYPE zif_abapgit_definitions=>ty_objects_tt,
-      END OF ty_push_result .
+    TYPES: BEGIN OF ty_pull_result,
+             files   TYPE zif_abapgit_definitions=>ty_files_tt,
+             objects TYPE zif_abapgit_definitions=>ty_objects_tt,
+             branch  TYPE zif_abapgit_definitions=>ty_sha1,
+             commit  TYPE zif_abapgit_definitions=>ty_sha1,
```
Comment (larshp) [inline src/git/zcl_abapgit_git_porcelain.clas.abap]: the branch and commit both is a commit sha1, so both are not needed at the same time?

File: src/git/zcl_abapgit_git_porcelain.clas.abap [hunk 3]
```diff
@@ -4,31 +4,32 @@ CLASS zcl_abapgit_git_porcelain DEFINITION
...
+    TYPES: BEGIN OF ty_push_result,
+             new_files     TYPE zif_abapgit_definitions=>ty_files_tt,
+             branch        TYPE zif_abapgit_definitions=>ty_sha1,
+             updated_files TYPE zif_abapgit_definitions=>ty_file_signatures_tt,
+             new_objects   TYPE zif_abapgit_definitions=>ty_objects_tt,
+           END OF ty_push_result .
 
     CLASS-METHODS pull
       IMPORTING
+        !io_repo         TYPE REF TO zcl_abapgit_repo
```
Comment (larshp) [inline src/git/zcl_abapgit_git_porcelain.clas.abap]: I'd like to keep the repo out of the git package (/src/git/), done multiple refactorings to get to the current state

File: src/git/zcl_abapgit_git_porcelain.clas.abap [hunk 4]
```diff
@@ -383,17 +381,26 @@ CLASS ZCL_ABAPGIT_GIT_PORCELAIN IMPLEMENTATION.
...
+    DATA: lv_sha1   TYPE zif_abapgit_definitions=>ty_sha1,
+          ls_object LIKE LINE OF it_objects,
           ls_commit TYPE zcl_abapgit_git_pack=>ty_commit.
 
+    IF    iv_commit IS SUPPLIED
+      AND iv_commit IS NOT INITIAL.
+      lv_sha1 = iv_commit.
+    ELSE.
+      lv_sha1 = iv_branch.
+    ENDIF.
```
Comment (larshp) [inline src/git/zcl_abapgit_git_porcelain.clas.abap]: so iv_commit and iv_branch does the same thing

File: src/git/zcl_abapgit_git_transport.clas.abap [hunk 5]
```diff
@@ -7,17 +7,19 @@ CLASS zcl_abapgit_git_transport DEFINITION
...
-        !ev_branch      TYPE zif_abapgit_definitions=>ty_sha1
-        !eo_branch_list TYPE REF TO zcl_abapgit_git_branch_list
-      RAISING
-        zcx_abapgit_exception .
+      IMPORTING io_repo         TYPE REF TO zcl_abapgit_repo OPTIONAL
+                iv_url          TYPE string
+                iv_branch_name  TYPE string
+                iv_commit_hash  TYPE zif_abapgit_definitions=>ty_sha1 OPTIONAL
+                iv_deepen       TYPE abap_bool DEFAULT abap_true
+                iv_deepen_level TYPE numc2 DEFAULT '1'
```
Comment (larshp) [inline src/git/zcl_abapgit_git_transport.clas.abap]: both parameters iv_deepen and iv_deepen_level are not needed, suggest if deepen = 0 then deepen is not enabled

File: src/git/zcl_abapgit_git_transport.clas.abap [hunk 6]
```diff
@@ -7,17 +7,19 @@ CLASS zcl_abapgit_git_transport DEFINITION
...
-        !iv_branch_name TYPE string
-        !iv_deepen      TYPE abap_bool DEFAULT abap_true
-        !it_branches    TYPE zif_abapgit_definitions=>ty_git_branch_list_tt OPTIONAL
-      EXPORTING
-        !et_objects     TYPE zif_abapgit_definitions=>ty_objects_tt
-        !ev_branch      TYPE zif_abapgit_definitions=>ty_sha1
-        !eo_branch_list TYPE REF TO zcl_abapgit_git_branch_list
-      RAISING
-        zcx_abapgit_exception .
+      IMPORTING io_repo         TYPE REF TO zcl_abapgit_repo OPTIONAL
```
Comment (larshp) [inline src/git/zcl_abapgit_git_transport.clas.abap]: repo...

File: src/git/zcl_abapgit_git_transport.clas.abap [hunk 7]
```diff
@@ -7,17 +7,19 @@ CLASS zcl_abapgit_git_transport DEFINITION
...
+                iv_url          TYPE string
+                iv_branch_name  TYPE string
+                iv_commit_hash  TYPE zif_abapgit_definitions=>ty_sha1 OPTIONAL
+                iv_deepen       TYPE abap_bool DEFAULT abap_true
+                iv_deepen_level TYPE numc2 DEFAULT '1'
+                it_branches     TYPE zif_abapgit_definitions=>ty_git_branch_list_tt OPTIONAL
+      EXPORTING et_objects      TYPE zif_abapgit_definitions=>ty_objects_tt
+                ev_branch       TYPE zif_abapgit_definitions=>ty_sha1
+                eo_branch_list  TYPE REF TO zcl_abapgit_git_branch_list
+                ev_commit       TYPE zif_abapgit_definitions=>ty_sha1
```
Comment (larshp) [inline src/git/zcl_abapgit_git_transport.clas.abap]: ev_branch and ev_commit contains the same value?

File: src/git/zcl_abapgit_git_transport.clas.abap [hunk 8]
```diff
@@ -290,37 +308,59 @@ CLASS ZCL_ABAPGIT_GIT_TRANSPORT IMPLEMENTATION.
...
-      zcx_abapgit_exception=>raise( 'empty pack' ).
-    ENDIF.
+        IF lv_pack IS INITIAL.
+          zcx_abapgit_exception=>raise( 'Response could not be parsed - empty pack returned.' ).
+        ENDIF.
+
+        et_objects = zcl_abapgit_git_pack=>decode( lv_pack ).
+
+      CATCH zcx_abapgit_exception INTO lx_exception.
+        IF   io_repo IS BOUND
```
Comment (larshp) [inline src/git/zcl_abapgit_git_transport.clas.abap]: this seems suspicious, git does not return an error that can be used?

File: src/zcl_abapgit_repo.clas.abap [hunk 9]
```diff
@@ -96,6 +95,14 @@ CLASS zcl_abapgit_repo DEFINITION
         !is_settings TYPE zif_abapgit_persistence=>ty_repo-local_settings
       RAISING
         zcx_abapgit_exception .
+    METHODS get_local_checksums
+      RETURNING
+        VALUE(rt_checksums) TYPE zif_abapgit_persistence=>ty_local_checksum_tt .
+    METHODS set_local_checksums
```
Comment (larshp) [inline src/zcl_abapgit_repo.clas.abap]: can we wait moving scope / introducing set_local_checksums until it is needed? the repo class is too big already, been doing a lot of refactoring to have less public methods

File: src/git/zcl_abapgit_git_porcelain.clas.abap (hunk 2 above)
Comment (mariusraht2) [reply src/git/zcl_abapgit_git_porcelain.clas.abap]: Yes, but we need it to separate a branch-related commit from a "normal" one:
grafik

File: src/git/zcl_abapgit_git_porcelain.clas.abap (hunk 1 above)
Comment (mariusraht2) [reply src/git/zcl_abapgit_git_porcelain.clas.abap]: Yes, will change it.

File: src/git/zcl_abapgit_git_porcelain.clas.abap (hunk 4 above)
Comment (mariusraht2) [reply src/git/zcl_abapgit_git_porcelain.clas.abap]: s. comment above

File: src/git/zcl_abapgit_git_transport.clas.abap (hunk 7 above)
Comment (mariusraht2) [reply src/git/zcl_abapgit_git_transport.clas.abap]: No, ev_branch contains latest commit and ev_commit contains the one selected by the user

File: src/git/zcl_abapgit_git_transport.clas.abap (hunk 8 above)
Comment (mariusraht2) [reply src/git/zcl_abapgit_git_transport.clas.abap]: Obsolete - removed reference to repo completely.

**Comment by mariusraht2** [comment]
Just fyi: Will wait with the second PR til this was successfully reviewed.

**Comment by larshp** [comment]
hi,
please remove changes in
```
zif_abapgit_persistence.intf.abap
zcl_abapgit_gui_chunk_lib.clas.abap
zcl_abapgit_merge.clas.abap
zcl_abapgit_repo.clas.abap 
zcl_abapgit_repo_online.clas.abap
zif_abapgit_definitions.intf.abap
```
unless its strictly needed due to refactorings in `git_porcelain` and `git_transport`

I cannot keep track of it all at once

**Comment by mariusraht2** [comment]
Oops..yes, didn't wanted that - sry.

**Comment by mariusraht2** [comment]
Most of them are already related to each other: 
- In zif_abapgit_persistence the data types for the repo attribute setter mask is being defined
- In zcl_abapgit_repo and zcl_abapgit_repo_online is the definition/implementation for the setter and getter of the commit hash
- In zcl_abapgit_merge I had to adjust the call of zcl_abapgit_git_transport=>upload_pack
- In zif_abapgit_definitions I added some definitions for commits for zcl_abapgit_git_transport

I was able to remove the changes in zcl_abapgit_gui_chunk_lib though and removed the changes related to beautification in zif_abapgit_definitions.

**Comment by mariusraht2** [comment]
@larshp Is there anything else I could do or are you still reviewing? Just asking, not trying to put you under pressure :)

**Comment by larshp** [comment]
So, first method changed is `pull`, instead of adding a parameter to the method, suggest making a new method `pull_by_commit`, perhaps rename `pull` to `pull_by_branch` but this can be done in the future in a follow up PR.

Method `push`, is this needed for "Checkout Commit" ?

**Comment by mariusraht2** [comment]
> Method `push`, is this needed for "Checkout Commit" ?

Yes, to get the full list of objects to be pushed back to the Git Platform I have to determine the objects by the commit and not branch hash. So I separated these different fields and due to that `full_tree` is being called with the commit hash in `push`.

**Comment by larshp** [comment]
okay, so pushing is in scope for this PR? Please help updating the description

**Comment by mariusraht2** [comment]
Updated description accordingly to your question.

**Comment by larshp** [comment]
okay, so I suggest closing this PR and opening a new with the _sole purpose_ to 

1: rename `pull` to `pull_by_branch`
2: create new method `pull_by_commit`

**Comment by mariusraht2** [comment]
How should we proceed with push and the other changes?

**Comment by larshp** [comment]
in another incremental change after the first change

**Comment by mariusraht2** [comment]
But first we would have to change zcl_abapgit_git_transport=>upload_pack. Should I open a new PR for this too?

**Comment by larshp** [comment]
yes please

**Comment by larshp** [comment]
closing

**Pull request closed without merging**